### PR TITLE
Tasty Reader: unpickle sealed children more lazily

### LINKED ISSUE #12369: scala3 integration: illegal cyclic inheritance involving object
## reproduction steps

using Scala (2.13.5+, 3.0.0-x),

Test.scala
```scala
package test

sealed trait A

object A {
  sealed trait B
  case object First extends A with B
  case object Second extends A with B
}
```

Main.scala

```scala
import test.A

object Main extends App {
}
```

build.sbt
```scala
lazy val root = project.in(file(".")).aggregate(project3, project213)

lazy val project3 = project.in(file("pr3")).settings(Seq(scalaVersion := "3.0.0-RC1"))

lazy val project213 = project.in(file("pr213")).dependsOn(project3).settings(Seq(
  resolvers += "scala-integration" at "https://scala-ci.typesafe.com/artifactory/scala-integration/",
  scalaVersion := "2.13.6-bin-fd69ef8",
  scalacOptions ++= (if (isDotty.value) Seq.empty else Seq("-Ytasty-reader"))
))
```

## problem

Doesn't compile when module with version 2.13x depends on module with 3.0.0x, but compiles when both modules have are the same versions.

### PR DESCRIPTION
**Pull request by bishabosha**
Also adds more detailed tracing

fixes scala/bug#12369

### CODE REVIEW

**Comment by bishabosha** [comment]
I will be on holiday until 20th July - after that I plan to also address https://github.com/scala/bug/issues/12420

File: src/compiler/scala/tools/nsc/tasty/ForceKinds.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  /** When forcing the companion of a module */
+  final val DeepForce: ForceKinds.Single = of(1 << 1)
+  /** When forcing the owner of a symbol */
+  final val CompleteOwner: ForceKinds.Single = of(1 << 2)
+  /** When forcing an overloaded signature */
+  final val OverloadedSym: ForceKinds.Single = of(1 << 3)
+  /** When forcing a symbol that will be copied */
+  final val CopySym: ForceKinds.Single = of(1 << 4)
+  /** When forcing the underlying symbol of some type space */
+  final val SpaceForce: ForceKinds.Single = of(1 << 5)
```
Comment (SethTisue) [inline src/compiler/scala/tools/nsc/tasty/ForceKinds.scala:32]: 🇺🇸 Seal_of_the_United_States_Space_Force svg

**Review by dwijnand — APPROVED** [review]
LGTM, interesting tracing framework - is it from elsewhere?

File: src/compiler/scala/tools/nsc/tasty/ForceKinds.scala (hunk 1 above)
Comment (bishabosha) [reply src/compiler/scala/tools/nsc/tasty/ForceKinds.scala:32]: I couldn't resist 😂

**Comment by lrytz** [comment]
👍

**Comment by bishabosha** [comment]
> LGTM, interesting tracing framework - is it from elsewhere?

my main motivation was that there are a lot of operations that cause deeply nested forcing of symbols, and I wanted to see the eventual result that they yield, without the pain of scanning around the logs to find them - so each "task" has a unique Id that can be used to jump between the query and result, (it also makes it visually simple to see nesting.)

I would say its inspired at least by the tracing in Dotty - which has a similar "query/result" structure

**Pull request merged**